File: core/src/main/java/org/iumotionlabs/hefesto/feature/portinfo/model/EnrichedPortBinding.java
package org.iumotionlabs.hefesto.feature.portinfo.model;

import java.util.Optional;

/**
 * Record representing a port binding enriched with additional context.
 */
public record EnrichedPortBinding(
    PortBinding binding,
    ServiceInfo serviceInfo,
    ProcessInfo processInfo,
    DockerInfo dockerInfo
) {
    /**
     * Creates an enriched binding from a basic binding.
     */
    public static EnrichedPortBinding from(PortBinding binding) {
        return new EnrichedPortBinding(binding, null, null, null);
    }

    /**
     * Creates an enriched binding with service info.
     */
    public static EnrichedPortBinding withService(PortBinding binding, ServiceInfo serviceInfo) {
        return new EnrichedPortBinding(binding, serviceInfo, null, null);
    }

    /**
     * Returns a copy with service info.
     */
    public EnrichedPortBinding withServiceInfo(ServiceInfo info) {
        return new EnrichedPortBinding(binding, info, processInfo, dockerInfo);
    }

    /**
     * Returns a copy with process info.
     */
    public EnrichedPortBinding withProcessInfo(ProcessInfo info) {
        return new EnrichedPortBinding(binding, serviceInfo, info, dockerInfo);
    }

    /**
     * Returns a copy with Docker info.
     */
    public EnrichedPortBinding withDockerInfo(DockerInfo info) {
        return new EnrichedPortBinding(binding, serviceInfo, processInfo, info);
    }

    /**
     * Returns optional service info.
     */
    public Optional<ServiceInfo> serviceInfoOpt() {
        return Optional.ofNullable(serviceInfo);
    }

    /**
     * Returns optional process info.
     */
    public Optional<ProcessInfo> processInfoOpt() {
        return Optional.ofNullable(processInfo);
    }

    /**
     * Returns optional Docker info.
     */
    public Optional<DockerInfo> dockerInfoOpt() {
        return Optional.ofNullable(dockerInfo);
    }

    // Delegate methods to binding
    public int port() {
        return binding.port();
    }

    public String protocol() {
        return binding.protocol();
    }

    public String state() {
        return binding.state();
    }

    public long pid() {
        return binding.pid();
    }

    public String processName() {
        return binding.processName();
    }

    public String localAddress() {
        return binding.localAddress();
    }

    public String user() {
        return binding.user();
    }

    /**
     * Checks if the port is exposed to the network (0.0.0.0 or ::).
     */
    public boolean isExposed() {
        String addr = binding.localAddress();
        return "0.0.0.0".equals(addr) || "::".equals(addr) || "*".equals(addr);
    }

    /**
     * Checks if bound to localhost only.
     */
    public boolean isLocalOnly() {
        String addr = binding.localAddress();
        return "127.0.0.1".equals(addr) || "::1".equals(addr) || "localhost".equals(addr);
    }

    /**
     * Checks if this is a Docker container.
     */
    public boolean isDocker() {
        return dockerInfo != null;
    }

    /**
     * Returns a formatted text representation.
     */
    public String toText() {
        StringBuilder sb = new StringBuilder();
        sb.append(binding.toText());

        if (serviceInfo != null) {
            sb.append(" ").append(serviceInfo.toTag());
        }

        return sb.toString();
    }

    /**
     * Returns a compact representation with service tag.
     */
    public String toCompactWithService() {
        String base = binding.toCompact();
        if (serviceInfo != null) {
            return base + " " + serviceInfo.toTag();
        }
        return base;
    }
}

File: hefesto-domain/src/portinfo/enriched_port_binding.rs
use serde::Serialize;

use super::docker_info::DockerInfo;
use super::port_binding::{ConnectionState, PortBinding, Protocol};
use super::process_info::ProcessInfo;
use super::service_info::ServiceInfo;

/// A port binding enriched with additional context.
#[derive(Debug, Clone, Serialize)]
pub struct EnrichedPortBinding {
    pub binding: PortBinding,
    pub service_info: Option<ServiceInfo>,
    pub process_info: Option<ProcessInfo>,
    pub docker_info: Option<DockerInfo>,
}

impl EnrichedPortBinding {
    /// Creates an enriched binding from a basic binding.
    pub fn from_binding(binding: PortBinding) -> Self {
        Self {
            binding,
            service_info: None,
            process_info: None,
            docker_info: None,
        }
    }

    /// Creates an enriched binding with service info.
    pub fn with_service(binding: PortBinding, service_info: ServiceInfo) -> Self {
        Self {
            binding,
            service_info: Some(service_info),
            process_info: None,
            docker_info: None,
        }
    }

    /// Returns a copy with service info.
    pub fn set_service_info(mut self, info: ServiceInfo) -> Self {
        self.service_info = Some(info);
        self
    }

    /// Returns a copy with process info.
    pub fn set_process_info(mut self, info: ProcessInfo) -> Self {
        self.process_info = Some(info);
        self
    }

    /// Returns a copy with Docker info.
    pub fn set_docker_info(mut self, info: DockerInfo) -> Self {
        self.docker_info = Some(info);
        self
    }

    // Delegate methods to binding
    pub fn port(&self) -> u16 {
        self.binding.port
    }

    pub fn protocol(&self) -> &Protocol {
        &self.binding.protocol
    }

    pub fn state(&self) -> &ConnectionState {
        &self.binding.state
    }

    pub fn pid(&self) -> u32 {
        self.binding.pid
    }

    pub fn process_name(&self) -> &str {
        &self.binding.process_name
    }

    pub fn local_address(&self) -> &str {
        &self.binding.local_address
    }

    pub fn user(&self) -> &str {
        &self.binding.user
    }

    /// Checks if the port is exposed to the network (0.0.0.0 or ::).
    pub fn is_exposed(&self) -> bool {
        let addr = &self.binding.local_address;
        addr == "0.0.0.0" || addr == "::" || addr == "*"
    }

    /// Checks if bound to localhost only.
    pub fn is_local_only(&self) -> bool {
        let addr = &self.binding.local_address;
        addr == "127.0.0.1" || addr == "::1" || addr == "localhost"
    }

    /// Checks if this is a Docker container.
    pub fn is_docker(&self) -> bool {
        self.docker_info.is_some()
    }

    /// Returns a formatted text representation.
    pub fn to_text(&self) -> String {
        let mut sb = self.binding.to_text();
        if let Some(ref si) = self.service_info {
            sb.push(' ');
            sb.push_str(&si.to_tag());
        }
        sb
    }

    /// Returns a compact representation with service tag.
    pub fn to_compact_with_service(&self) -> String {
        let base = self.binding.to_compact();
        if let Some(ref si) = self.service_info {
            format!("{} {}", base, si.to_tag())
        } else {
            base
        }
    }
}
